### PR TITLE
arrays-arrays_with_methods-hard-q9

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [x] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        Object output = MethodUtil.invokeIfMethodExists(ArrayofDigits.class, "powerN", new Object[]{given}, int[].class);
+        assertNull(output, "Your powerN method should not have a return variable.");
+        CustomAssertions._assertArrayEquals(expected, given, "Your method does not square the values in the given array.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("mainMethodInputProvider")
+    void correctMainMethodOutput(String in, int one, int two, int three, int four, int five) {
+        runWithInput(in);
+        assertEquals(one, Integer.parseInt(getItemByName("first")), "Your program does not square the first number.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:78]: All the error messages should be "... square the x number in the array."

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Comment by SamamthaH** [comment]
Do you want me to change up the test so that the length of the array can be varied in the website?

File: src/arrays/arrays_with_methods/hard/q9/ArrayofDigits.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package arrays.arrays_with_methods.hard.q9;
+
+/*
+Write a Java program that asks the user to enter 5 integers to be stored in an integer array. Next, define a method called
+powerN that takes an integer array, computes the 2nd power of each element, and does not return any output. Back in the main
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/ArrayofDigits.java:5]:
```suggestion
powerN that takes an integer array and squares each element. The return type of powerN should be void, and it should be generalized to handle any input array length. Back in the main
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    @ParameterizedTest
+    @MethodSource("powerNInputProvider")
+    void correctPowerNMethod(int[] given) throws Throwable {
+        int[] expected = ans(given);
+        Object[][] arguments = {
+                {given, int[].class}
+        };
+        MethodTest m = new MethodTest(ArrayofDigits.class, "powerN", arguments);
+        Object output = m.callMethod();
+        assertNull(output, "Your powerN method should not have a return variable.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:87]:
```suggestion
        assertNull(output, "Your powerN does not have the correct return type.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        CustomAssertions._assertArrayEquals(expected, given, "Your method does not square the values in the given array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:88]:
```suggestion
        CustomAssertions._assertArrayEquals(expected, given, "Your powerN method does not square the values in the input array.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("mainMethodInputProvider")
+    void correctMainMethodOutput(String in, int one, int two, int three, int four, int five) {
+        runWithInput(in);
+        assertEquals(one, Integer.parseInt(getItemByName("first")), "Your program does not square the first number in the array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:95]:
```suggestion
        assertEquals(one, Integer.parseInt(getItemByName("first")), "Your program does not correctly square and print the array elements.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        assertEquals(two, Integer.parseInt(getItemByName("second")), "Your program does not square the first number in the array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:96]:
```suggestion
        assertEquals(two, Integer.parseInt(getItemByName("second")), "Your program does not correctly square and print the array elements.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        assertEquals(three, Integer.parseInt(getItemByName("third")), "Your program does not square the first number in the array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:97]:
```suggestion
        assertEquals(three, Integer.parseInt(getItemByName("third")), "Your program does not correctly square and print the array elements.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        assertEquals(four, Integer.parseInt(getItemByName("fourth")), "Your program does not square the first number in the array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:98]:
```suggestion
        assertEquals(four, Integer.parseInt(getItemByName("fourth")), "Your program does not correctly square and print the array elements.");
```

File: src/arrays/arrays_with_methods/hard/q9/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        assertEquals(five, Integer.parseInt(getItemByName("fifth")), "Your program does not square the first number in the array.");
```
Comment (Hedgemon4) [inline src/arrays/arrays_with_methods/hard/q9/MainTest.java:99]:
```suggestion
        assertEquals(five, Integer.parseInt(getItemByName("fifth")), "Your program does not correctly square and print the array elements.");
```

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**